File: 2023-2024/sourcecode/mod2/lab9/SharedPreferencesBasicsJava/app/src/main/java/com/example/sharedpreferencesbasicsjava/ui/UserFragment.java
package com.example.sharedpreferencesbasicsjava.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.sharedpreferencesbasicsjava.data.dummydata.DataProvider;
import com.example.sharedpreferencesbasicsjava.databinding.FragmentUserBinding;
import com.example.sharedpreferencesbasicsjava.viewmodel.UserViewModel;

public class UserFragment extends Fragment {

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final FragmentUserBinding binding = FragmentUserBinding.inflate(inflater);

        UserViewModel viewModel =new ViewModelProvider(this).get(UserViewModel.class);
        viewModel.getUsername().observe(requireActivity(), binding.textView::setText);

        binding.addButton.setOnClickListener(v -> viewModel.addUser(DataProvider.getUsername()));

        binding.clearButton.setOnClickListener(v -> viewModel.clearUsers());

        return binding.getRoot();
    }
}

File: 2023-2024/sourcecode/mod2/lab9/RoomBasicsKotlin/app/src/main/java/com/example/roombasicskotlin/ui/UserFragment.kt
package com.example.roombasicskotlin.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.roombasicskotlin.data.dummydata.DataProvider
import com.example.roombasicskotlin.databinding.FragmentUserBinding
import com.example.roombasicskotlin.viewmodel.UserViewModel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class UserFragment : Fragment() {

    private lateinit var binding: FragmentUserBinding

    private val viewModel: UserViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentUserBinding.inflate(inflater)

        val userAdapter = UserAdapter(UserComparator())
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.usersState.collectLatest { users ->
                userAdapter.submitList(users)
            }
        }

        binding.rvList.apply{
            adapter = userAdapter
            layoutManager = LinearLayoutManager(requireContext())
            itemAnimator = null
        }

        binding.addButton.setOnClickListener { viewModel.addUser(DataProvider.user) }
        binding.clearButton.setOnClickListener { viewModel.clearUsers() }

        return binding.root
    }
}